deps/rabbit/src/rabbit_stream_sac_coordinator.erl

```diff
@@ -0,0 +1,629 @@
+                               consumers = Consumers0} =
+                            Group) ->
+    ActiveConsumerIndex = PartitionIndex rem length(Consumers0),
+    rabbit_log:debug("Active consumer index = ~p rem ~p = ~p",
+                     [PartitionIndex, length(Consumers0), ActiveConsumerIndex]),
+    {_, Consumers1} =
+        lists:foldr(fun (C0, {Index, Cs}) when Index == ActiveConsumerIndex ->
+                            C1 = C0#consumer{active = true},
+                            {Index - 1, [C1 | Cs]};
+                        (C0, {Index, Cs}) ->
+                            C1 = C0#consumer{active = false},
+                            {Index - 1, [C1 | Cs]}
+                    end,
+                    {length(Consumers0) - 1, []}, Consumers0),
+    Group#group{consumers = Consumers1}.
+
+evaluate_active_consumer(#group{partition_index = PartitionIndex,
+                                consumers = Consumers}) ->
+    ActiveConsumerIndex = PartitionIndex rem length(Consumers),
+    lists:nth(ActiveConsumerIndex + 1, Consumers).
+
+lookup_consumer(ConnectionPid, SubscriptionId,
+                #group{consumers = Consumers}) ->
+    lists:search(fun(#consumer{pid = ConnPid, subscription_id = SubId}) ->
+                    ConnPid == ConnectionPid andalso SubId == SubscriptionId
+                 end,
+                 Consumers).
+
+lookup_active_consumer(#group{consumers = Consumers}) ->
+    lists:search(fun(#consumer{active = Active}) -> Active end,
+                 Consumers).
+
+update_groups(VirtualHost,
+              Stream,
+              ConsumerName,
+              #group{consumers = []},
+              StreamGroups) ->
+    rabbit_log:debug("Group ~p ~p ~p is now empty, removing it",
+                     [VirtualHost, Stream, ConsumerName]),
+    maps:remove({VirtualHost, Stream, ConsumerName}, StreamGroups);
+update_groups(VirtualHost,
+              Stream,
+              ConsumerName,
+              Group,
+              StreamGroups) ->
+    maps:put({VirtualHost, Stream, ConsumerName}, Group, StreamGroups).
+
+update_consumer_state_in_group(#group{consumers = Consumers0} = G,
+                               Pid,
+                               SubId,
+                               NewState) ->
+    CS1 = lists:foldr(fun(C0, Acc) ->
+                         case C0 of
+                             #consumer{pid = Pid, subscription_id = SubId} ->
+                                 C1 = C0#consumer{active = NewState},
+                                 [C1 | Acc];
+                             C -> [C | Acc]
+                         end
+                      end,
+                      [], Consumers0),
+    G#group{consumers = CS1}.
+
+mod_call_effect(Pid, Msg) ->
+    {mod_call, rabbit_stream_sac_coordinator, send_message, [Pid, Msg]}.
+
+-spec send_message(pid(), term()) -> ok.
+send_message(ConnectionPid, Msg) ->
+    ConnectionPid ! Msg,
+    ok.
```

Comment:
Why `mod_call` effect instead of directly `send_msg` effect?

Reply:
`mod_call` provides more guarantee of execution than `send_msg` IIRC. Is that correct @kjnilsson?